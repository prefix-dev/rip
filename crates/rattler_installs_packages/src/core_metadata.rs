// Implementation comes from https://github.com/njsmith/posy/blob/main/src/vocab/core_metadata.rs
// Licensed under MIT or Apache-2.0

use crate::extra::Extra;
use crate::package_name::PackageName;
use crate::requirement::PackageRequirement;
use crate::rfc822ish::RFC822ish;
use crate::specifier::Specifiers;
use once_cell::sync::Lazy;
use pep440::Version;
use std::collections::HashSet;

#[derive(Debug, Clone)]
pub struct WheelCoreMetadata {
    pub name: PackageName,
    pub version: Version,
    pub requires_dist: Vec<PackageRequirement>,
    pub requires_python: Specifiers,
    pub extras: HashSet<Extra>,
}

impl TryFrom<&[u8]> for WheelCoreMetadata {
    type Error = miette::Report;

    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
        let (name, version, mut parsed) = parse_common(value)?;

        let mut requires_dist = Vec::new();
        for req_str in parsed.take_all("Requires-Dist").into_iter() {
            match req_str.parse() {
                Err(e) => {
                    tracing::error!("ignoring Requires-Dist: {req_str}, failed to parse: {e}")
                }
                Ok(req) => requires_dist.push(req),
            }
        }

        let requires_python = match parsed.maybe_take("Requires-Python")? {
            Some(rp_str) => rp_str.parse()?,
            None => Specifiers(Vec::new()),
        };

        let mut extras: HashSet<Extra> = HashSet::new();
        for extra in parsed.take_all("Provides-Extra").drain(..) {
            extras.insert(extra.parse()?);
        }

        Ok(WheelCoreMetadata {
            name,
            version,
            requires_dist,
            requires_python,
            extras,
        })
    }
}

fn parse_common(input: &[u8]) -> miette::Result<(PackageName, Version, RFC822ish)> {
    let input = String::from_utf8_lossy(input);
    let mut parsed = RFC822ish::parse(&input)?;

    static NEXT_MAJOR_METADATA_VERSION: Lazy<Version> = Lazy::new(|| Version::parse("3").unwrap());

    // Quoth https://packaging.python.org/specifications/core-metadata:
    // "Automated tools consuming metadata SHOULD warn if metadata_version
    // is greater than the highest version they support, and MUST fail if
    // metadata_version has a greater major version than the highest
    // version they support (as described in PEP 440, the major version is
    // the value before the first dot)."
    //
    // We do the MUST, but I think I disagree about warning on
    // unrecognized minor revisions. If it's a minor revision, then by
    // definition old software is supposed to be able to handle it "well
    // enough". The only purpose of the warning would be to alert users
    // that they might want to upgrade, or to alert the tool authors that
    // there's a new metadata release. But for users, there are better
    // ways to nudge them to upgrade (e.g. checking on startup, like
    // pip does), and new metadata releases are so rare and so
    // much-discussed beforehand that if a tool's authors don't know
    // about it it's because the tool is abandoned anyway.
    let metadata_version = parsed.take("Metadata-Version")?;
    let metadata_version: Version = Version::parse(&metadata_version)
        .ok_or_else(|| miette::miette!("failed to parse {metadata_version}"))?;
    if metadata_version >= *NEXT_MAJOR_METADATA_VERSION {
        miette::bail!("unsupported Metadata-Version {}", metadata_version);
    }

    let version_str = parsed.take("Version")?;

    Ok((
        parsed.take("Name")?.parse()?,
        Version::parse(&version_str)
            .ok_or_else(|| miette::miette!("failed to parse version '{version_str}'"))?,
        parsed,
    ))
}
