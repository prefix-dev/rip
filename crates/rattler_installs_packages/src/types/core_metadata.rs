// Implementation comes from https://github.com/njsmith/posy/blob/main/src/vocab/core_metadata.rs
// Licensed under MIT or Apache-2.0

use super::extra::ParseExtraError;
use crate::{
    types::Extra, types::PackageName, types::ParsePackageNameError, types::RFC822ish,
    types::Version, types::VersionSpecifiers,
};
use once_cell::sync::Lazy;
use pep440_rs::Pep440Error;
use pep508_rs::Requirement;
use std::{collections::HashSet, str::FromStr};
use thiserror::Error;

#[derive(Debug, Clone)]

/// The core metadata of a wheel.
pub struct WheelCoreMetadata {
    /// The name of the package
    pub name: PackageName,
    /// Version w.r.t to PEP440
    pub version: Version,
    /// Version of the metadata
    pub metadata_version: MetadataVersion,
    /// Requirements for this distribution
    /// Matches the Requires-Dist field
    pub requires_dist: Vec<Requirement>,
    /// Python requirement
    pub requires_python: Option<VersionSpecifiers>,
    /// Extras provided by this distribution
    pub extras: HashSet<Extra>,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
/// Wrapper around a PEP440 version
/// specifically for the metadata version
pub struct MetadataVersion(pub Version);

impl MetadataVersion {
    /// We consider that this implements PEP643
    /// if the version is 2.3 or higher.
    pub fn implements_pep643(&self) -> bool {
        static VERSION_2_2: Lazy<MetadataVersion> = Lazy::new(|| {
            MetadataVersion(Version::from_str("2.2").expect("cannot parse 2.2 version string"))
        });

        if self < &VERSION_2_2 {
            return false;
        }
        true
    }
}

#[derive(Debug, Error)]
#[allow(missing_docs)]
pub enum WheelCoreMetaDataError {
    #[error(transparent)]
    FailedToParseMetadata(#[from] <RFC822ish as FromStr>::Err),

    #[error("missing key {0} in METADATA")]
    MissingKey(String),

    #[error("duplicate key {0} in METADATA")]
    DuplicateKey(String),

    #[error("invalid Metadata-Version: {0}")]
    InvalidMetadataVersion(String),

    #[error("invalid Version: {0}")]
    InvalidVersion(String),

    #[error("invalid Requires-Python: {0}")]
    InvalidRequiresPython(#[source] Pep440Error),

    #[error("unsupported METADATA version {0}")]
    UnsupportedVersion(Version),

    #[error(transparent)]
    InvalidPackageName(#[from] ParsePackageNameError),

    #[error("invalid extra identifier '{0}'")]
    InvalidExtra(String, #[source] ParseExtraError),

    #[error("{0}")]
    FailedToParse(String),
}

impl TryFrom<&[u8]> for WheelCoreMetadata {
    type Error = WheelCoreMetaDataError;

    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
        let (name, version, metadata_version, mut parsed) = parse_common(value)?;

        let mut requires_dist = Vec::new();
        for req_str in parsed.take_all("Requires-Dist").into_iter() {
            match req_str.parse() {
                Err(e) => {
                    tracing::warn!("ignoring Requires-Dist: {req_str}, failed to parse: {e}")
                }
                Ok(req) => requires_dist.push(req),
            }
        }

        let requires_python = parsed
            .maybe_take("Requires-Python")
            .map_err(|_| WheelCoreMetaDataError::DuplicateKey(String::from("Requires-Python")))?
            .as_deref()
            .map(VersionSpecifiers::from_str)
            .transpose()
            .map_err(WheelCoreMetaDataError::InvalidRequiresPython)?;

        let mut extras: HashSet<Extra> = HashSet::new();
        for extra in parsed.take_all("Provides-Extra").drain(..) {
            extras.insert(
                extra
                    .parse()
                    .map_err(|e| WheelCoreMetaDataError::InvalidExtra(extra, e))?,
            );
        }

        Ok(WheelCoreMetadata {
            name,
            version,
            metadata_version,
            requires_dist,
            requires_python,
            extras,
        })
    }
}

fn parse_common(
    input: &[u8],
) -> Result<(PackageName, Version, MetadataVersion, RFC822ish), WheelCoreMetaDataError> {
    let input = String::from_utf8_lossy(input);
    let mut parsed = RFC822ish::from_str(&input)?;

    static NEXT_MAJOR_METADATA_VERSION: Lazy<Version> =
        Lazy::new(|| Version::from_str("3").unwrap());

    // Quoth https://packaging.python.org/specifications/core-metadata:
    // "Automated tools consuming metadata SHOULD warn if metadata_version
    // is greater than the highest version they support, and MUST fail if
    // metadata_version has a greater major version than the highest
    // version they support (as described in PEP 440, the major version is
    // the value before the first dot)."
    //
    // We do the MUST, but I think I disagree about warning on
    // unrecognized minor revisions. If it's a minor revision, then by
    // definition old software is supposed to be able to handle it "well
    // enough". The only purpose of the warning would be to alert users
    // that they might want to upgrade, or to alert the tool authors that
    // there's a new metadata release. But for users, there are better
    // ways to nudge them to upgrade (e.g. checking on startup, like
    // pip does), and new metadata releases are so rare and so
    // much-discussed beforehand that if a tool's authors don't know
    // about it it's because the tool is abandoned anyway.
    let metadata_version = parsed
        .take("Metadata-Version")
        .map_err(|_| WheelCoreMetaDataError::MissingKey(String::from("Metadata-Version")))?;
    let metadata_version: Version = metadata_version
        .parse()
        .map_err(WheelCoreMetaDataError::InvalidMetadataVersion)?;
    if metadata_version >= *NEXT_MAJOR_METADATA_VERSION {
        return Err(WheelCoreMetaDataError::UnsupportedVersion(metadata_version));
    }

    let version_str = parsed
        .take("Version")
        .map_err(|_| WheelCoreMetaDataError::MissingKey(String::from("Version")))?;

    Ok((
        parsed
            .take("Name")
            .map_err(|_| WheelCoreMetaDataError::MissingKey(String::from("Name")))?
            .parse()?,
        version_str
            .parse()
            .map_err(WheelCoreMetaDataError::InvalidVersion)?,
        MetadataVersion(metadata_version),
        parsed,
    ))
}
