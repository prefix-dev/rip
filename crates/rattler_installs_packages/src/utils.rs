use futures::{AsyncRead, AsyncReadExt, AsyncSeekExt};
use include_dir::{include_dir, Dir};
use std::io::{Read, Seek, SeekFrom};
use tokio_util::compat::TokioAsyncReadCompatExt;
use url::Url;

/// Keep retrying a certain IO function until it either succeeds or until it doesn't return
/// [`std::io::ErrorKind::Interrupted`].
pub fn retry_interrupted<F, T>(mut f: F) -> std::io::Result<T>
where
    F: FnMut() -> std::io::Result<T>,
{
    loop {
        match f() {
            Ok(result) => return Ok(result),
            Err(err) if err.kind() != std::io::ErrorKind::Interrupted => {
                return Err(err);
            }
            _ => {
                // Otherwise keep looping!
            }
        }
    }
}

/// Represents either data coming from the network in an async fashion or a local thing on disk.
/// We only use async for the network stuff, the local filesystem doesn't really benefit from it.
pub enum StreamingOrLocal {
    Streaming(Box<dyn AsyncRead + Unpin + Send>),
    Local(Box<dyn ReadAndSeek + Send>),
}

pub trait ReadAndSeek: Read + Seek {}
impl<T> ReadAndSeek for T where T: Read + Seek {}

impl StreamingOrLocal {
    /// Returns an instance that is both readable and seekable by first streaming the contents to
    /// disk if required.
    pub async fn force_local(self) -> std::io::Result<Box<dyn ReadAndSeek + Send>> {
        Ok(match self {
            StreamingOrLocal::Local(stream) => stream,
            StreamingOrLocal::Streaming(mut stream) => {
                let mut tmp = tokio::fs::File::from(tempfile::tempfile()?).compat();
                futures::io::copy(&mut stream, &mut tmp).await?;
                tmp.seek(SeekFrom::Start(0)).await?;
                Box::new(tmp.into_inner().into_std().await)
            }
        })
    }

    pub async fn read_to_end(self, bytes: &mut Vec<u8>) -> std::io::Result<usize> {
        match self {
            StreamingOrLocal::Streaming(mut streaming) => streaming.read_to_end(bytes).await,
            StreamingOrLocal::Local(mut local) => {
                match tokio::task::spawn_blocking(move || {
                    let mut bytes = Vec::new();
                    local.read_to_end(&mut bytes).map(|_| bytes)
                })
                .await
                {
                    Ok(Ok(result)) => {
                        *bytes = result;
                        Ok(bytes.len())
                    }
                    Ok(Err(err)) => Err(err),
                    Err(err) => {
                        if let Ok(panic) = err.try_into_panic() {
                            std::panic::resume_unwind(panic)
                        }
                        Err(std::io::ErrorKind::Interrupted.into())
                    }
                }
            }
        }
    }
}

/// Normalize url according to pip standards
pub fn normalize_index_url(mut url: Url) -> Url {
    let path = url.path();
    if !path.ends_with('/') {
        url.set_path(&format!("{path}/"));
    }
    url
}

pub(crate) static VENDORED_PACKAGING_DIR: Dir<'_> =
    include_dir!("$CARGO_MANIFEST_DIR/vendor/packaging/");
