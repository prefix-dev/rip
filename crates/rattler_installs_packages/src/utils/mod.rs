mod read_and_seek;
mod streaming_or_local;

mod seek_slice;

use include_dir::{include_dir, Dir};
use url::Url;

pub use read_and_seek::ReadAndSeek;
pub use streaming_or_local::StreamingOrLocal;

pub use seek_slice::SeekSlice;

/// Keep retrying a certain IO function until it either succeeds or until it doesn't return
/// [`std::io::ErrorKind::Interrupted`].
pub fn retry_interrupted<F, T>(mut f: F) -> std::io::Result<T>
where
    F: FnMut() -> std::io::Result<T>,
{
    loop {
        match f() {
            Ok(result) => return Ok(result),
            Err(err) if err.kind() != std::io::ErrorKind::Interrupted => {
                return Err(err);
            }
            _ => {
                // Otherwise keep looping!
            }
        }
    }
}

/// Normalize url according to pip standards
pub fn normalize_index_url(mut url: Url) -> Url {
    let path = url.path();
    if !path.ends_with('/') {
        url.set_path(&format!("{path}/"));
    }
    url
}

pub(crate) static VENDORED_PACKAGING_DIR: Dir<'_> =
    include_dir!("$CARGO_MANIFEST_DIR/vendor/packaging/");
