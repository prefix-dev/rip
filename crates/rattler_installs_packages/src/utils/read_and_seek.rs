use std::io::{Read, Seek};

/// Defines that a type can be read and seeked. This trait has a blanket implementation for any type
/// that implements both [`Read`] and [`Seek`].
pub trait ReadAndSeek: Read + Seek {}

impl<T> ReadAndSeek for T where T: Read + Seek {}
