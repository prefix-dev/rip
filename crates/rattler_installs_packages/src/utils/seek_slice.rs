// Implementation comes from https://github.com/njsmith/posy/blob/main/src/seek_slice.rs
// Licensed under MIT or Apache-2.0

use std::io;
use std::io::{Read, Seek, SeekFrom};

pub struct SeekSlice<T> {
    inner: T,
    start: u64,
    end: u64,
    current: u64,
}

impl<T: Seek> SeekSlice<T> {
    pub fn new(mut inner: T, start: u64, end: u64) -> io::Result<SeekSlice<T>> {
        assert!(end >= start);
        // initialize current position to something sensible
        let current = inner.seek(SeekFrom::Start(start))?;
        Ok(SeekSlice {
            inner,
            start,
            end,
            current,
        })
    }
}

impl<T: Seek> Seek for SeekSlice<T> {
    fn seek(&mut self, pos: SeekFrom) -> io::Result<u64> {
        let maybe_goal_idx = match pos {
            SeekFrom::Start(amount) => self.start.checked_add(amount),
            SeekFrom::End(amount) => self.end.checked_add_signed(amount),
            SeekFrom::Current(amount) => self.current.checked_add_signed(amount),
        };
        match maybe_goal_idx {
            Some(goal_idx) => {
                if goal_idx < self.start || goal_idx > self.end {
                    Err(io::Error::new(
                        io::ErrorKind::InvalidInput,
                        format!("invalid seek to a negative or overflowing position (goal: {}, start: {}, end: {})", goal_idx, self.start, self.end),
                    ))
                } else {
                    self.current = self.inner.seek(SeekFrom::Start(goal_idx))?;
                    Ok(self.current.checked_sub(self.start).unwrap())
                }
            }
            None => Err(io::Error::new(
                io::ErrorKind::InvalidInput,
                "integer overflow while seeking",
            )),
        }
    }
}

impl<T: Read + Seek> Read for SeekSlice<T> {
    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
        let max_read: usize = (self.end - self.current).try_into().unwrap_or(usize::MAX);
        let read_size = std::cmp::min(max_read, buf.len());
        let amount = self.inner.read(&mut buf[..read_size])?;
        self.current += amount as u64;
        Ok(amount)
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use std::io::Cursor;

    #[test]
    fn test_seek_slice() {
        let buf: &[u8] = &[0, 1, 2, 3, 4, 5, 6, 7, 8, 9];
        let mut cursor = Cursor::new(&buf);
        let mut slice = SeekSlice::new(&mut cursor, 2, 8).unwrap();
        // starts at offset zero
        assert_eq!(slice.stream_position().unwrap(), 0);
        // reading advances position as expected
        fn next_byte<T: Read>(value: T) -> u8 {
            value.bytes().next().unwrap().unwrap()
        }
        assert_eq!(next_byte(&mut slice), 2u8);
        assert_eq!(next_byte(&mut slice), 3u8);
        assert_eq!(slice.stream_position().unwrap(), 2);
        assert_eq!(next_byte(&mut slice), 4u8);

        // out of range seeks caught and have no effect
        assert!(slice.seek(SeekFrom::Current(-10)).is_err());
        assert!(slice.seek(SeekFrom::Current(10)).is_err());
        assert_eq!(next_byte(&mut slice), 5u8);

        assert_eq!(slice.seek(SeekFrom::Start(1)).unwrap(), 1);
        assert_eq!(next_byte(&mut slice), 3u8);

        assert_eq!(slice.seek(SeekFrom::End(-1)).unwrap(), 5);
        assert_eq!(next_byte(&mut slice), 7u8);
        assert!(slice.bytes().next().is_none());
    }
}
