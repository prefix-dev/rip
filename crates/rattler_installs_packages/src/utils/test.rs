use std::sync::Arc;

use reqwest::Client;
use reqwest_middleware::ClientWithMiddleware;
use tempfile::TempDir;

use crate::{
    index::{PackageDb, PackageSourcesBuilder},
    python_env::Pep508EnvMakers,
    resolve::solve_options::ResolveOptions,
    wheel_builder::WheelBuilder,
};

pub fn get_package_db() -> (Arc<PackageDb>, TempDir) {
    let tempdir = tempfile::tempdir().unwrap();
    let client = ClientWithMiddleware::from(Client::new());

    let url = url::Url::parse("https://pypi.org/simple/").unwrap();
    let sources = PackageSourcesBuilder::new(url).build().unwrap();

    (
        Arc::new(PackageDb::new(sources, client, tempdir.path(), Default::default()).unwrap()),
        tempdir,
    )
}

// Setup the test environment
pub async fn setup(resolve_options: ResolveOptions) -> (Arc<WheelBuilder>, TempDir) {
    let (package_db, tempdir) = get_package_db();
    let env_markers = Arc::new(Pep508EnvMakers::from_env().await.unwrap().0);

    (
        WheelBuilder::new(
            package_db.clone(),
            env_markers.clone(),
            None,
            resolve_options,
        )
        .unwrap(),
        tempdir,
    )
}
