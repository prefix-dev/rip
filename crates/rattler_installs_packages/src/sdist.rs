use crate::core_metadata::WheelCoreMetadata;
use crate::utils::ReadAndSeek;
use crate::{Artifact, MetadataArtifact, SDistFormat, SDistName};
use flate2::read::GzDecoder;
use miette::{miette, IntoDiagnostic};
use parking_lot::Mutex;
use std::ffi::OsStr;
use std::io::Read;
use std::path::Path;
use std::str::FromStr;
use tar::Archive;

/// Represents a source distribution artifact.
pub struct SDist {
    /// Name of the source distribution
    name: SDistName,

    /// Source dist archive
    archive: Mutex<Archive<Box<dyn Read + Send>>>,
}

impl SDist {
    /// Create this struct from a path
    #[allow(dead_code)]
    pub fn from_path(path: &Path) -> miette::Result<Self> {
        let file_name = path
            .file_name()
            .and_then(OsStr::to_str)
            .ok_or_else(|| miette::miette!("path does not contain a filename"))?;
        let name = SDistName::from_str(file_name).into_diagnostic()?;
        let bytes = std::fs::File::open(path).into_diagnostic()?;
        Self::new(name, Box::new(bytes))
    }

    /// Find entry in tar archive
    fn find_entry(&self, name: impl AsRef<str>) -> miette::Result<Option<Vec<u8>>> {
        let mut archive = self.archive.lock();
        // Loop over entries
        for entry in archive.entries().into_diagnostic()? {
            let mut entry = entry.into_diagnostic()?;

            // Find name in archive and return this
            if entry.path().into_diagnostic()?.ends_with(name.as_ref()) {
                let mut bytes = Vec::new();
                entry.read_to_end(&mut bytes).into_diagnostic()?;
                return Ok(Some(bytes));
            }
        }
        Ok(None)
    }

    /// Read .PKG-INFO from the archive
    pub fn read_package_info(&self) -> miette::Result<(Vec<u8>, WheelCoreMetadata)> {
        if let Some(bytes) = self.find_entry("PKG-INFO")? {
            let metadata = Self::parse_metadata(&bytes)?;
            Ok((bytes, metadata))
        } else {
            Err(miette!("no PKG-INFO found in archive"))
        }
    }

    /// Read the build system info from the pyproject.toml
    #[allow(dead_code)]
    pub fn read_build_info(&self) -> miette::Result<pyproject_toml::BuildSystem> {
        if let Some(bytes) = self.find_entry("pyproject.toml")? {
            let source = String::from_utf8(bytes).into_diagnostic()?;
            let project = pyproject_toml::PyProjectToml::new(&source).into_diagnostic()?;
            Ok(project
                .build_system
                .ok_or_else(|| miette!("no build-system found in pyproject.toml"))?)
        } else {
            Err(miette!("no pyproject.toml found in archive"))
        }
    }
}

impl Artifact for SDist {
    type Name = SDistName;

    fn new(name: Self::Name, bytes: Box<dyn ReadAndSeek + Send>) -> miette::Result<Self> {
        let sdist = match name.format {
            SDistFormat::TarGz => {
                let bytes = Box::new(GzDecoder::new(bytes));
                Self {
                    name,
                    archive: Mutex::new(Archive::new(bytes)),
                }
            }
            SDistFormat::Tar => {
                let bytes: Box<dyn Read + Send> = Box::new(bytes);
                Self {
                    name,
                    archive: Mutex::new(Archive::new(bytes)),
                }
            }
            _ => return Err(miette!("unsupported format {:?}", name.format)),
        };
        Ok(sdist)
    }

    fn name(&self) -> &Self::Name {
        &self.name
    }
}

/// We can re-use the metadata type from the wheel if the SDist has a PKG-INFO.
type SDistMetadata = WheelCoreMetadata;

impl MetadataArtifact for SDist {
    type Metadata = SDistMetadata;

    fn parse_metadata(bytes: &[u8]) -> miette::Result<Self::Metadata> {
        WheelCoreMetadata::try_from(bytes).into_diagnostic()
    }

    fn metadata(&self) -> miette::Result<(Vec<u8>, Self::Metadata)> {
        // Assume we have a PKG-INFO
        self.read_package_info()
    }
}

#[cfg(test)]
mod tests {
    use crate::MetadataArtifact;
    use insta::{assert_debug_snapshot, assert_ron_snapshot};
    use std::path::Path;

    #[test]
    pub fn read_rich_metadata() {
        // Read path
        let path =
            Path::new(env!("CARGO_MANIFEST_DIR")).join("../../test-data/sdists/rich-13.6.0.tar.gz");

        // Load sdist
        let sdist = super::SDist::from_path(&path).unwrap();

        let metadata = sdist.metadata().unwrap().1;
        assert_debug_snapshot!(metadata);
    }

    #[test]
    pub fn read_rich_build_info() {
        // Read path
        let path =
            Path::new(env!("CARGO_MANIFEST_DIR")).join("../../test-data/sdists/rich-13.6.0.tar.gz");

        // Load sdist
        let sdist = super::SDist::from_path(&path).unwrap();

        let build_system = sdist.read_build_info().unwrap();

        assert_ron_snapshot!(build_system, @r###"
        BuildSystem(
          requires: [
            "poetry-core >=1.0.0",
          ],
          r#build-backend: Some("poetry.core.masonry.api"),
          r#backend-path: None,
        )
        "###);
    }
}
