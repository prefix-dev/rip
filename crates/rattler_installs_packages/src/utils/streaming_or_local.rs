use crate::utils::ReadAndSeek;
use futures::TryFutureExt;
use std::{
    io,
    io::{Read, Seek, Write},
};
use tempfile::SpooledTempFile;
use tokio::io::{AsyncRead, AsyncReadExt};
use tokio::task::JoinError;

/// Represents a stream of data that is either coming in asynchronously from a remote source or from
/// a synchronous location (like the filesystem).
///
/// It is often useful to make this distinction because reading from a remote source is often slower
/// than reading synchronously (from disk or memory).
pub enum StreamingOrLocal {
    /// Represents an asynchronous stream of data.
    Streaming(Box<dyn AsyncRead + Unpin + Send>),

    /// Represents a synchronous stream of data.
    Local(Box<dyn ReadAndSeek + Send>),
}

impl StreamingOrLocal {
    /// Stream in the contents of the stream and make sure we have a fast locally accessible stream.
    ///
    /// If the stream is already local this will simply return that stream. If however the file is
    /// remote it will first be read to a temporary spooled file.
    pub async fn into_local(self) -> io::Result<Box<dyn ReadAndSeek + Send>> {
        match self {
            StreamingOrLocal::Streaming(mut stream) => {
                // Create a [`SpooledTempFile`] which is a blob of memory that is kept in memory if
                // it does not grow beyond 5MB, otherwise it is written to disk.
                let mut local_file = SpooledTempFile::new(5 * 1024 * 1024);

                // Stream in the bytes and copy them to the temporary file.
                let mut buf = [0u8; 1024 * 8];
                loop {
                    let bytes_read = stream.read(&mut buf).await?;
                    if bytes_read == 0 {
                        break;
                    }
                    local_file.write_all(&buf)?;
                }

                // Restart the file from the start so we can start reading from it.
                local_file.rewind()?;

                Ok(Box::new(local_file))
            }
            StreamingOrLocal::Local(stream) => Ok(stream),
        }
    }

    /// Asynchronously read the contents of the stream into a vector of bytes.
    pub async fn read_to_end(self, bytes: &mut Vec<u8>) -> std::io::Result<usize> {
        match self {
            StreamingOrLocal::Streaming(mut streaming) => streaming.read_to_end(bytes).await,
            StreamingOrLocal::Local(mut local) => {
                match tokio::task::spawn_blocking(move || {
                    let mut bytes = Vec::new();
                    local.read_to_end(&mut bytes).map(|_| bytes)
                })
                .map_err(JoinError::try_into_panic)
                .await
                {
                    Ok(Ok(result)) => {
                        *bytes = result;
                        Ok(bytes.len())
                    }
                    Ok(Err(err)) => Err(err),
                    // Resume the panic on the main task
                    Err(Ok(panic)) => std::panic::resume_unwind(panic),
                    Err(Err(_)) => Err(std::io::ErrorKind::Interrupted.into()),
                }
            }
        }
    }
}
