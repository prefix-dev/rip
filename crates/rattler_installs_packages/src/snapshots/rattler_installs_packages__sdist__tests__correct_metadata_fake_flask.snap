---
source: crates/rattler_installs_packages/src/sdist.rs
expression: metadata
---
WheelCoreMetadata {
    name: PackageName {
        source: "Flask",
        normalized: "flask",
    },
    version: Version {
        epoch: 0,
        release: [
            3,
            0,
            0,
        ],
        pre: None,
        post: None,
        dev: None,
        local: None,
    },
    metadata_version: MetadataVersion(
        Version {
            epoch: 0,
            release: [
                2,
                2,
            ],
            pre: None,
            post: None,
            dev: None,
            local: None,
        },
    ),
    requires_dist: [
        Requirement {
            name: "Werkzeug",
            extras: None,
            version_or_url: Some(
                VersionSpecifier(
                    VersionSpecifiers(
                        [
                            VersionSpecifier {
                                operator: GreaterThanEqual,
                                version: Version {
                                    epoch: 0,
                                    release: [
                                        3,
                                        0,
                                        0,
                                    ],
                                    pre: None,
                                    post: None,
                                    dev: None,
                                    local: None,
                                },
                            },
                        ],
                    ),
                ),
            ),
            marker: None,
        },
        Requirement {
            name: "Jinja2",
            extras: None,
            version_or_url: Some(
                VersionSpecifier(
                    VersionSpecifiers(
                        [
                            VersionSpecifier {
                                operator: GreaterThanEqual,
                                version: Version {
                                    epoch: 0,
                                    release: [
                                        3,
                                        1,
                                        2,
                                    ],
                                    pre: None,
                                    post: None,
                                    dev: None,
                                    local: None,
                                },
                            },
                        ],
                    ),
                ),
            ),
            marker: None,
        },
        Requirement {
            name: "itsdangerous",
            extras: None,
            version_or_url: Some(
                VersionSpecifier(
                    VersionSpecifiers(
                        [
                            VersionSpecifier {
                                operator: GreaterThanEqual,
                                version: Version {
                                    epoch: 0,
                                    release: [
                                        2,
                                        1,
                                        2,
                                    ],
                                    pre: None,
                                    post: None,
                                    dev: None,
                                    local: None,
                                },
                            },
                        ],
                    ),
                ),
            ),
            marker: None,
        },
        Requirement {
            name: "click",
            extras: None,
            version_or_url: Some(
                VersionSpecifier(
                    VersionSpecifiers(
                        [
                            VersionSpecifier {
                                operator: GreaterThanEqual,
                                version: Version {
                                    epoch: 0,
                                    release: [
                                        8,
                                        1,
                                        3,
                                    ],
                                    pre: None,
                                    post: None,
                                    dev: None,
                                    local: None,
                                },
                            },
                        ],
                    ),
                ),
            ),
            marker: None,
        },
        Requirement {
            name: "blinker",
            extras: None,
            version_or_url: Some(
                VersionSpecifier(
                    VersionSpecifiers(
                        [
                            VersionSpecifier {
                                operator: GreaterThanEqual,
                                version: Version {
                                    epoch: 0,
                                    release: [
                                        1,
                                        6,
                                        2,
                                    ],
                                    pre: None,
                                    post: None,
                                    dev: None,
                                    local: None,
                                },
                            },
                        ],
                    ),
                ),
            ),
            marker: None,
        },
        Requirement {
            name: "importlib-metadata",
            extras: None,
            version_or_url: Some(
                VersionSpecifier(
                    VersionSpecifiers(
                        [
                            VersionSpecifier {
                                operator: GreaterThanEqual,
                                version: Version {
                                    epoch: 0,
                                    release: [
                                        3,
                                        6,
                                        0,
                                    ],
                                    pre: None,
                                    post: None,
                                    dev: None,
                                    local: None,
                                },
                            },
                        ],
                    ),
                ),
            ),
            marker: Some(
                Expression(
                    MarkerExpression {
                        l_value: MarkerEnvVersion(
                            PythonVersion,
                        ),
                        operator: LessThan,
                        r_value: QuotedString(
                            "3.10",
                        ),
                    },
                ),
            ),
        },
        Requirement {
            name: "asgiref",
            extras: None,
            version_or_url: Some(
                VersionSpecifier(
                    VersionSpecifiers(
                        [
                            VersionSpecifier {
                                operator: GreaterThanEqual,
                                version: Version {
                                    epoch: 0,
                                    release: [
                                        3,
                                        2,
                                    ],
                                    pre: None,
                                    post: None,
                                    dev: None,
                                    local: None,
                                },
                            },
                        ],
                    ),
                ),
            ),
            marker: Some(
                Expression(
                    MarkerExpression {
                        l_value: Extra,
                        operator: Equal,
                        r_value: QuotedString(
                            "async",
                        ),
                    },
                ),
            ),
        },
        Requirement {
            name: "python-dotenv",
            extras: None,
            version_or_url: None,
            marker: Some(
                Expression(
                    MarkerExpression {
                        l_value: Extra,
                        operator: Equal,
                        r_value: QuotedString(
                            "dotenv",
                        ),
                    },
                ),
            ),
        },
    ],
    requires_python: Some(
        VersionSpecifiers(
            [
                VersionSpecifier {
                    operator: GreaterThanEqual,
                    version: Version {
                        epoch: 0,
                        release: [
                            3,
                            8,
                        ],
                        pre: None,
                        post: None,
                        dev: None,
                        local: None,
                    },
                },
            ],
        ),
    ),
    extras: {
        Extra {
            source: "dotenv",
            normalized: "dotenv",
        },
        Extra {
            source: "async",
            normalized: "async",
        },
    },
}
